dependencies.yaml

```diff
@@ -76,6 +75,16 @@ dependencies:
         packages:
           - wheel
           - setuptools>=61.0.0
+    specific:
+      - output_types: [conda, requirements, pyproject]
+        matrices:
+          - matrix:
+              cuda: "12.0"
+            packages:
+              - &cuda_python12 cuda-python>=12.1,<13.0
```

Comment:
@vyasr I don't think that this requirement is being seen by CI. [It appears to pull `cuda-python` 11.8.1.](https://github.com/rapidsai/rmm/actions/runs/4875577291/jobs/8698100831?pr=1259#step:10:104) I think this is because the pip dependencies aren't rendered out at build time like conda dependencies and we don't have a good way to express CUDA-specific dependencies in `pyproject.toml`. How should we address that?

Reply:
In the interest of expediency, I think the best approach in the short term is to add some logic to https://github.com/rapidsai/rmm/blob/branch-23.06/ci/release/apply_wheel_modifications.sh that will modify the cuda-python version in the pyproject.toml file prior to the build based on the current CUDA version. In the longer term we may need to move back to using a custom build backend, but I would like to avoid that complexity for now. It will require maintaining the dependency in one extra place, but I think that's OK since we've making similar expedient decisions in the conda recipes for CUDA 12.

Reply:
I would accept that. In my opinion, we should do that modification by rendering with `rapids-dependency-file-generator`, but I believe that isn't supported yet. I tried `rapids-dependency-file-generator --file_key py_build --matrix "cuda=12.0"` but it needs an `--output` and `pyproject` isn't listed. I can't remember where our previous issue/discussion about this is located.

Reply:
I would prefer using dfg for this as well. We can revisit as an enhancement to dfg once we have working CUDA 12 wheels.